Owncloud iOs Client/InstantUpload/InstantUpload.m

```diff
@@ -0,0 +1,242 @@
+//
+//  InstantUpload.h
+//  Owncloud iOs Client
+//
+
+/*
+ Copyright (C) 2016, ownCloud, Inc.
+ This code is covered by the GNU Public License Version 3.
+ For distribution utilizing Apple mechanisms please see https://owncloud.org/contribute/iOS-license-exception/
+ You should have received a copy of this license
+ along with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.en.html>.
+*/
+
+#import <CoreLocation/CoreLocation.h>
+#import <Photos/Photos.h>
+
+#import "InstantUpload.h"
+
+#import "AppDelegate.h"
+#import "constants.h"
+
+#import "ManageAppSettingsDB.h"
+#import "ManageUsersDB.h"
+#import "UserDto.h"
+
+#define ACTIVE_USER ((AppDelegate *)[[UIApplication sharedApplication] delegate]).activeUser
+
+@interface InstantUpload () <CLLocationManagerDelegate, PHPhotoLibraryChangeObserver>
+
+@property (strong, nonatomic) CLLocationManager *locationManager;
+
+@end
+
+@implementation InstantUpload
+
++ (instancetype) instantUploadManager {
+    static dispatch_once_t once;
+    static InstantUpload *__instantUploadManager;
+    dispatch_once(&once, ^{
+        __instantUploadManager = [[InstantUpload alloc] init];
+    });
+    return __instantUploadManager;
+}
+
+- (instancetype) init {
+    if (self = [super init]) {
+        self.locationManager = [[CLLocationManager alloc] init];
+        self.locationManager.delegate = self;
+        [PHPhotoLibrary.sharedPhotoLibrary registerChangeObserver:self];
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appLifecycleShouldTriggerInstantUpload:) name:UIApplicationWillEnterForegroundNotification object:nil];
+    }
+    return self;
+}
+
+- (void) dealloc {
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (BOOL) enabled {
+    return ACTIVE_USER.instantUpload;
+}
+
+- (BOOL) backgroundInstantUploadEnabled {
+    return ACTIVE_USER.backgroundInstantUpload;
+}
+
+- (void) setEnabled:(BOOL)enabled {
+    @synchronized (self) {
+        if (enabled != self.enabled) {
+            if (enabled) {
+                if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusAuthorized) {
+                    ACTIVE_USER.instantUpload = YES;
+                    [ManageAppSettingsDB updateInstantUploadTo:YES];
+                    [self attemptUpload];
```

Comment:
So that images taken before instant upload is enabled aren't uploaded? I believe the behavior of Dropbox and similar apps (Google Photos) is to upload your entire history, but I'm not sure what OwnCloud has done in the past.